hdc/algo/ops/stats.py

```diff
@@ -0,0 +1,423 @@
+
+
+@numba.njit
+def gammafit(x):
+    # pylint: disable=line-too-long
+    """
+    Calculate gamma distribution parameters for timeseries.
+
+    Adapted from:
+    https://github.com/scipy/scipy/blob/f2ef65dc7f00672496d7de6154744fee55ef95e9/scipy/stats/_continuous_distns.py#L2554
+    Copyright (c) 2001-2002 Enthought, Inc.  2003-2019, SciPy Developers.
+    All rights reserved.
+    """
+    n = 0
+    xts = 0
+    logs = 0
+
+    for xx in x:
+        if xx > 0:
+            xts += xx
+            logs += log(xx)
+            n += 1
+
+    if n == 0:
+        return (0, 0)
+
+    xtsbar = xts / n
+    s = log(xtsbar) - (logs / n)
+
+    if s == 0:
+        return (0, 0)
+
+    a_est = (3 - s + sqrt((s - 3) ** 2 + 24 * s)) / (12 * s)
+    xa = a_est * (1 - 0.4)
+    xb = a_est * (1 + 0.4)
+    a = brentq(xa, xb, s)
+    if a == 0:
+        return (0, 0)
+
+    b = xtsbar / a
+
+    return (a, b)
+
+
+@numba.njit
+def spifun(x, a=None, b=None, cal_start=None, cal_stop=None):
+    """Calculate SPI with gamma distribution for 3d array."""
+    y = np.full_like(x, -9999)
+    r, c, t = x.shape
+
+    if cal_start is None:
+        cal_start = 0
+
+    if cal_stop is None:
+        cal_stop = t
+
+    cal_ix = np.arange(cal_start, cal_stop)
+
+    for ri in range(r):
+        for ci in range(c):
+            xt = x[ri, ci, :]
+            valid_ix = []
+
+            for tix in range(t):
+                if xt[tix] > 0:
+                    valid_ix.append(tix)
+
+            n_valid = len(valid_ix)
+
+            p_zero = 1 - (n_valid / t)
+
+            if p_zero > 0.9:
+                y[ri, ci, :] = -9999
+                continue
+
+            if a is None or b is None:
+                alpha, beta = gammafit(xt[cal_ix])
+            else:
+                alpha, beta = (a, b)
+
+            if alpha == 0 or beta == 0:
+                y[ri, ci, :] = -9999
+                continue
+
+            spi = np.full(t, p_zero, dtype=nt.float64)  # type: ignore
+
+            for tix in valid_ix:
+                spi[tix] = p_zero + (
+                    (1 - p_zero)
+                    * sc.gammainc(alpha, xt[tix] / beta)  # pylint: disable=no-member
+                )
+
+            for tix in range(t):
+                spi[tix] = sc.ndtri(spi[tix]) * 1000
+
+            np.round_(spi, 0, spi)
+
+            y[ri, ci, :] = spi[:]
+
+    return y
+
+
+@numba.njit
+def mk_score(x):
+    """Calculate MK score (S) and Kendall's Tau.
+
+    https://vsp.pnnl.gov/help/vsample/design_trend_mann_kendall.htm
+
+    Hussain et al., (2019). pyMannKendall:
+    a python package for non parametric Mann Kendall family of trend tests..
+    Journal of Open Source Software, 4(39), 1556, https://doi.org/10.21105/joss.01556
+    """
+    n = len(x)
+    _s1 = 0
+    _s2 = 0
+
+    for k in range(0, n - 1):
+        for kk in range(k + 1, n):
+            if x[kk] > x[k]:
+                _s1 += 1
+            if x[kk] < x[k]:
+                _s2 += 1
```

Comment:
Is the double loop more efficient in nopython jit than the numpy operations in the original pymankendall?
```
for k in range(n-1):
        s = s + np.sum(demo[k+1:n][x[k+1:n] > x[k]]) - np.sum(demo[k+1:n][x[k+1:n] < x[k]])
```        
If so, that's good to know!

Reply:
yes in general it's more efficient - although good to check on a by case basis